src/main/php/lang/ast/emit/Declaration.class.php

```diff
@@ -15,6 +15,43 @@ public function __construct($type) {
   /** @return string */
   public function name() { return ltrim($this->type->name, '\\'); }
 
+  /** @return iterable */
+  public function implementedInterfaces() {
+    if ($this->type instanceof InterfaceDeclaration) {
+      foreach ($this->type->parents as $interface) {
+        yield $interface->literal();
+      }
+    } else {
+      foreach ($this->type->implements as $interface) {
+        yield $interface->literal();
+      }
+    }
+  }
+
+  /**
+   * Checks whether a given method exists
+   *
+   * @param  string $named
+   * @return bool
+   */
+  public function providesMethod($named) {
+    return isset($this->body["{$named}()"]);
+  }
+
+  /**
+   * Returns all methods annotated with a given annotation
+   *
+   * @param  string $annotation
+   * @return iterable
+   */
+  public function methodsAnnotated($annotation) {
+    foreach ($this->body as $member) {
+      if ($member instanceof Method && $member->annotations && $member->annotations->named($annotation)) {
+        yield $member->name => $member->line;
```

Comment:
This is extremely closely modeled towards the current usecase. If we want to check modifiers, parameters, return type or anything else, we have no way of doing this. Maybe we should model this along `xp-framework/reflection` instead?